classes/engine.php

```diff
@@ -0,0 +1,456 @@
+        }
+
+        return $returnval;
+    }
+
+    /**
+     * Check if index exists in Elasticssearch backend
+     *
+     * @return bool True on success False on failure
+     */
+    private function check_index() {
+        $returnval = false;
+        $response = 404;
+        $url = $this->get_url();
+        $client = new \search_elastic\elastic_curl();
+
+        if (!empty($this->config->index) && $url) {
+            $index = $url . '/'. $this->config->index;
+            $client->get($index);
+            $response = $client->info['http_code'];
+        }
+        if ($response == 200) {
+            $returnval = true;
+        }
+
+        return $returnval;
+    }
+
+    /**
+     * Create index in Elasticsearch backend
+     */
+    private function create_index() {
+        $url = $this->get_url();
+        $client = new \search_elastic\elastic_curl();
+        if (!empty($this->config->index) && $url) {
+            $index = $url . '/'. $this->config->index;
+            $response = $client->put($index);
+        } else {
+            throw new \moodle_exception('noconfig', 'search_elastic', '');
+        }
+        if ($client->info['http_code'] !== 200) {
+            throw new \moodle_exception('indexfail', 'search_elastic', '');
+        }
+
+    }
+
+    /**
+     * Is the Elasticsearch server endpoint configured in Moodle
+     * and available.
+     *
+     * @return true|string Returns true if all good or an error string.
+     */
+    public function is_server_ready() {
+        $url = $this->get_url();
+        $returnval = true;
+        $client = new \search_elastic\elastic_curl();
+
+        if (!$url) {
+            $returnval = get_string('noconfig', 'search_elastic');
+        } else if (!(bool)json_decode($client->get($url))) {
+            $returnval = get_string('noserver', 'search_elastic');
+        }
+
+        return $returnval;
+    }
+
+    /**
+     * Called when indexing is triggered.
+     * Creates the Index namespace and adds fields if they don't exist.
+     */
+    public function index_starting($fullindex = false) {
+        // Check if index exists and create it if it doesn't.
+        $hasindex = $this->check_index();
+        if (!$hasindex) {
+            $this->create_index();
+        }
+    }
+
+    /**
+     * Return a count of total records for the most recently completed
+     * execute_query().
+     * Must be implemented to return the number of results that available
+     * for the most recent call to execute_query().
+     * This is used to determine how many pages will be displayed in the paging bar.
+     * For more discussion see MDL-53758.
+     *
+     * @return int
+     */
+    public function get_query_total_count() {
+        return $this->totalresultdocs;
+    }
+
+    /**
+     * Add a document to the index
+     */
+    public function add_document($document, $fileindexing = false) {
+        $docdata = $document->export_for_engine();
+        $url = $this->get_url();
```

Comment:
We can assume at this point that URL's are valid. The code will never get this far as an earlier check will catch it. However this whole pattern is a candidate for refactoring, which I'll look at